File: monitor.py
#!/usr/bin/env python3
"""Monitor Scentoria collections for newly added products and push a phone
notification via ntfy (https://ntfy.sh) for each new item.

State (the set of product IDs already seen) is stored as JSON under state/ and
committed back to the repo by the GitHub Action, so it persists between runs.

Required env var:
    NTFY_TOPIC   - your secret ntfy topic name (e.g. scentoria-xxxxxxxx)
Optional env vars:
    NTFY_SERVER  - ntfy server base URL (default: https://ntfy.sh)
    NTFY_TOKEN   - bearer token, only if you use a reserved/auth-protected topic
"""

import json
import os
import sys
import urllib.error
import urllib.request
from pathlib import Path

# Per-collection config:
#   max_pages - how much of the feed to read (None = the whole collection).
#   mode      - what counts as an alert:
#       "added"     -> a product ID newly appears in the feed (a new listing).
#       "restocked" -> a product becomes available: sold-out -> in stock, or a
#                      brand-new in-stock item. Needed because this store keeps
#                      sold-out items listed and just flips their stock flag, so a
#                      restock is NOT a new product ID — it's an availability change.
#
# new-arrivals is sorted newest-first and has no sold-out items, so we watch a
#   small window from the top of the feed for newly-listed products.
# restocked-gems is "Featured"-sorted and keeps ~half its items sold-out-but-listed,
#   so we read the whole collection and watch for flips to in-stock.
COLLECTIONS = {
    "new-arrivals": {
        "url": "https://scentoria.co.in/collections/new-arrivals/products.json",
        "max_pages": 3,      # top ~750 newest; far more headroom than needed per 5 min
        "mode": "added",
    },
    "restocked-gems": {
        "url": "https://scentoria.co.in/collections/restocked-gems/products.json",
        "max_pages": None,   # read the entire collection
        "mode": "restocked",
    },
}
PRODUCT_URL = "https://scentoria.co.in/products/{handle}"
COLLECTION_URL = "https://scentoria.co.in/collections/{collection}"

STATE_DIR = Path("state")
USER_AGENT = "scentoria-monitor/1.0 (+https://github.com/)"
PAGE_SIZE = 250          # Shopify max per page
HARD_PAGE_CAP = 25       # safety stop for full fetches (= up to 6250 products)
# If more than this many new products appear at once, send one summary instead
# of spamming an individual notification per product.
MAX_INDIVIDUAL = 8

NTFY_SERVER = os.environ.get("NTFY_SERVER", "https://ntfy.sh").rstrip("/")
NTFY_TOPIC = os.environ.get("NTFY_TOPIC", "").strip()
NTFY_TOKEN = os.environ.get("NTFY_TOKEN", "").strip()


def fetch_products(url, max_pages=None):
    """Fetch products from a collection feed, following Shopify pagination.

    max_pages=None reads the whole collection (up to HARD_PAGE_CAP pages);
    an integer reads only that many pages from the top of the feed.
    """
    limit = max_pages if max_pages is not None else HARD_PAGE_CAP
    products = []
    for page in range(1, limit + 1):
        full_url = f"{url}?limit={PAGE_SIZE}&page={page}"
        req = urllib.request.Request(full_url, headers={"User-Agent": USER_AGENT})
        with urllib.request.urlopen(req, timeout=30) as resp:
            batch = json.load(resp).get("products", [])
        if not batch:
            break
        products.extend(batch)
        if len(batch) < PAGE_SIZE:
            break
    return products


def load_seen(name):
    """Return the set of previously-seen product IDs, or None on first run."""
    f = STATE_DIR / f"{name}.json"
    if not f.exists():
        return None
    try:
        return set(json.loads(f.read_text()))
    except (json.JSONDecodeError, ValueError):
        # Corrupt state: treat as empty so we re-seed rather than crash.
        return set()


def save_seen(name, ids):
    STATE_DIR.mkdir(exist_ok=True)
    (STATE_DIR / f"{name}.json").write_text(json.dumps(sorted(ids)))


def product_price(p):
    variants = p.get("variants") or []
    if variants and variants[0].get("price"):
        return variants[0]["price"]
    return None


def product_image(p):
    images = p.get("images") or []
    if images and images[0].get("src"):
        return images[0]["src"]
    return None


def is_available(p):
    """True if any variant of the product is in stock."""
    return any(v.get("available") for v in (p.get("variants") or []))


def tracked_products(products, mode):
    """Map of product-id -> product for the items this collection alerts on.

    "restocked" mode tracks only in-stock products, so a sold-out item coming
    back (or a new in-stock item) shows up as a newly-tracked id. "added" mode
    tracks every listed product, so a new listing shows up as a new id.
    """
    if mode == "restocked":
        return {str(p["id"]): p for p in products if is_available(p)}
    return {str(p["id"]): p for p in products}


def send_ntfy(title, message, click=None, tags=None, priority=None, icon=None):
    """Publish a notification using ntfy's JSON format (UTF-8 safe)."""
    if not NTFY_TOPIC:
        print("ERROR: NTFY_TOPIC is not set; cannot send notification.", file=sys.stderr)
        return False
    payload = {"topic": NTFY_TOPIC, "title": title, "message": message}
    if click:
        payload["click"] = click
    if tags:
        payload["tags"] = tags
    if priority:
        payload["priority"] = priority
    if icon:
        payload["icon"] = icon

    data = json.dumps(payload).encode("utf-8")
    headers = {"Content-Type": "application/json", "User-Agent": USER_AGENT}
    if NTFY_TOKEN:
        headers["Authorization"] = f"Bearer {NTFY_TOKEN}"

    req = urllib.request.Request(NTFY_SERVER, data=data, headers=headers, method="POST")
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            return 200 <= resp.status < 300
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", errors="replace")
        print(f"ntfy HTTP {e.code}: {body}", file=sys.stderr)
    except urllib.error.URLError as e:
        print(f"ntfy network error: {e}", file=sys.stderr)
    return False


def notify(collection, mode, products):
    label = collection.replace("-", " ").title()
    tag = "fire" if mode == "restocked" else "shopping_bags"

    if len(products) > MAX_INDIVIDUAL:
        preview = ", ".join(p.get("title", "?") for p in products[:5])
        headline = (f"{len(products)} back in stock in {label}"
                    if mode == "restocked"
                    else f"{len(products)} new in {label}")
        send_ntfy(
            title=headline,
            message=f"{preview} … and {len(products) - 5} more.",
            click=COLLECTION_URL.format(collection=collection),
            tags=[tag],
            priority=4,
        )
        return

    for p in products:
        price = product_price(p)
        price_str = f"₹{price}" if price else None
        if mode == "restocked":
            head = "In stock now" + (f" · {price_str}" if price_str else "")
        else:
            head = price_str or "New product"
        send_ntfy(
            title=p.get("title", "Product"),
            message=f"{head}\nin {label}",
            click=PRODUCT_URL.format(handle=p.get("handle", "")),
            tags=[tag],
            priority=4,
            icon=product_image(p),
        )


def main():
    if not NTFY_TOPIC:
        sys.exit("NTFY_TOPIC env var is required.")

    exit_code = 0
    for name, cfg in COLLECTIONS.items():
        mode = cfg.get("mode", "added")
        label = name.replace("-", " ").title()
        try:
            products = fetch_products(cfg["url"], cfg.get("max_pages"))
        except Exception as e:  # network / parse errors: skip without touching state
            print(f"[{name}] fetch failed: {e}", file=sys.stderr)
            exit_code = 1
            continue

        if not products:
            # Empty response is almost certainly a transient glitch, not a real
            # "collection is now empty". Skip so we never wipe state or misfire.
            print(f"[{name}] returned 0 products — skipping this run.", file=sys.stderr)
            exit_code = 1
            continue

        tracked = tracked_products(products, mode)
        seen = load_seen(name)

        if seen is None:
            save_seen(name, tracked.keys())
            print(f"[{name}] first run — seeded {len(tracked)} (mode={mode}, no alerts).")
            if mode == "restocked":
                msg = (f"Watching {label} for restocks. {len(tracked)} items "
                       f"currently in stock — you'll be alerted when sold-out "
                       f"items come back or new in-stock ones drop.")
            else:
                msg = (f"Now tracking {len(tracked)} products in {label}. "
                       f"You'll be alerted on new additions.")
            send_ntfy(title=f"Monitoring started: {label}", message=msg, tags=["eyes"])
            continue

        # A "fresh" item is one tracked now but not in the previous snapshot:
        #   added mode     -> a newly-listed product.
        #   restocked mode -> in stock now but wasn't before (back in stock, or a
        #                     brand-new in-stock item). Sold-outs/removals just
        #                     drop out of the set and never alert.
        fresh = [tracked[pid] for pid in tracked if pid not in seen]
        if fresh:
            print(f"[{name}] {len(fresh)} alert(s) (mode={mode}):")
            for p in fresh:
                print(f"    + {p.get('title')} ({p.get('handle')})")
            notify(name, mode, fresh)
        else:
            print(f"[{name}] nothing new ({len(tracked)} tracked, mode={mode}).")

        save_seen(name, tracked.keys())

    return exit_code


if __name__ == "__main__":
    sys.exit(main())

File: src/monitor.js
// Stock monitor — core logic (runs on Node, driven by src/run.mjs from a
// GitHub Actions cron). Pure and storage-agnostic: callers inject an `env`
// with { STORE_BASE_URL, STATE: {get,put}, TG_BOT_TOKEN, TG_CHAT_ID }.
//
// Watches two Shopify collections and pushes a phone notification via a
// Telegram bot:
//   - new-arrivals   -> alert when a NEW product is listed.
//   - restocked-gems -> alert when a tester/retail/partial/set variant comes
//                       BACK IN STOCK (decants/samples/miniatures/roll-ons are
//                       ignored). Restocks are confirmed in real time via the
//                       storefront .js endpoint before alerting.
//
// Both collections are read in FULL every run (no top-pages window): products
// can be re-listed with an old created_at and land deep in the feed, so a
// windowed read would miss them. Running on GitHub Actions (no per-run CPU cap)
// makes parsing the whole catalog fine — what the Cloudflare Worker's 10ms-CPU
// free-tier limit could not do.
//
// State (the set of seen ids per collection) is read via env.STATE every run
// and WRITTEN only when the set changes. A product is recorded as "seen" only
// once its notification has actually been delivered — a failed send leaves it
// unseen so it is retried next run instead of being silently lost.
//
// Required: env.STORE_BASE_URL, env.TG_BOT_TOKEN, env.TG_CHAT_ID.
//
// Why Telegram and not ntfy: ntfy.sh meters its free quota per *IP* ("basis":
// "ip"), and shared egress IPs burn the daily quota via unrelated tenants so
// every send 429s. The Telegram Bot API is keyed by bot token + chat,
// independent of source IP, with no daily cap.

// Collection handles only; the store base URL comes from env (a secret), so the
// target site isn't hard-coded in this repo. Both read in full (maxPages null).
const COLLECTIONS = [
  { name: "new-arrivals", maxPages: null, mode: "added" }, // ~5k items; can land deep
  { name: "restocked-gems", maxPages: null, mode: "restocked" },
];

// Variant types we IGNORE for restock alerts (small try-size formats). Matched
// case-insensitively as a substring of the variant name. "DECAANT" covers a
// real typo present in the store's data.
const EXCLUDE_VARIANT_TYPES = [
  "DECANT", "DECAANT", "SAMPLE", "MINIATURE", "ROLL ON", "ROLL-ON", "ROLLON",
];

const PAGE_SIZE = 250;
const HARD_PAGE_CAP = 60; // safety stop for full fetches (~15k items)
const MAX_INDIVIDUAL = 8; // beyond this, send one summary instead of N pushes
const UA =
  "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 " +
  "(KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36";

// Store URLs are derived from env.STORE_BASE_URL (a secret) at runtime.
const storeBase = (env) => (env.STORE_BASE_URL || "").replace(/\/+$/, "");
const productUrl = (base, h) => `${base}/products/${h}`;
const collectionUrl = (base, c) => `${base}/collections/${c}`;

// ---------------------------------------------------------------------------
// Helpers

function titleCase(name) {
  return name.split("-").map((w) => w.charAt(0).toUpperCase() + w.slice(1)).join(" ");
}

function variantName(v) {
  return v.option1 || v.title || "?";
}

function isNotifiableVariant(v) {
  const name = variantName(v).toUpperCase();
  return !EXCLUDE_VARIANT_TYPES.some((t) => name.includes(t));
}

function productPrice(p) {
  const v = (p.variants || [])[0];
  return v && v.price ? v.price : null;
}

// Escape the small set of characters Telegram's HTML parse mode treats specially.
function esc(s) {
  return String(s).replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

function setsEqual(a, b) {
  if (a.size !== b.size) return false;
  for (const x of a) if (!b.has(x)) return false;
  return true;
}

async function fetchProducts(url, maxPages) {
  const limit = maxPages == null ? HARD_PAGE_CAP : maxPages;
  const products = [];
  let page = 1;
  for (; page <= limit; page++) {
    const r = await fetch(`${url}?limit=${PAGE_SIZE}&page=${page}`, {
      headers: { "User-Agent": UA },
    });
    if (!r.ok) throw new Error(`HTTP ${r.status} for ${url} page ${page}`);
    const batch = (await r.json()).products || [];
    if (batch.length === 0) break;
    products.push(...batch);
    if (batch.length < PAGE_SIZE) break; // last (partial) page reached the end
  }
  // We stopped on the page cap with a still-full last page -> likely truncated.
  if (page > limit && maxPages == null) {
    console.warn(`[fetch] hit HARD_PAGE_CAP=${limit} for ${url}; collection may be truncated — raise the cap.`);
  }
  return products;
}

// Real-time set of in-stock variant-id strings from the storefront .js endpoint
// (same data the live page uses). Returns null if the check itself failed, so
// the caller can retry next run rather than act on a guess.
async function liveAvailableVariants(base, handle) {
  try {
    const r = await fetch(`${productUrl(base, handle)}.js`, { headers: { "User-Agent": UA } });
    if (!r.ok) throw new Error(`HTTP ${r.status}`);
    const data = await r.json();
    return new Set((data.variants || []).filter((v) => v.available).map((v) => String(v.id)));
  } catch (e) {
    console.error(`live-check failed for ${handle}: ${e}`);
    return null;
  }
}

async function loadSeen(env, name) {
  const arr = await env.STATE.get(name, "json"); // null if missing (first run)
  return arr === null ? null : new Set(arr);
}

async function saveSeen(env, name, ids) {
  // Sorted so the committed state file shows minimal, readable diffs run-to-run.
  await env.STATE.put(name, JSON.stringify([...ids].sort()));
}

// Send one message via the Telegram Bot API. Returns true only if Telegram
// accepted it (HTTP 2xx) — callers rely on this to decide whether to record the
// item as seen. `preview` shows a rich link card (product photo) for the URL.
async function sendTelegram(env, { text, url, preview = false }) {
  if (!env.TG_BOT_TOKEN || !env.TG_CHAT_ID) {
    console.error("TG_BOT_TOKEN/TG_CHAT_ID not set; cannot notify.");
    return false;
  }
  const api = `https://api.telegram.org/bot${env.TG_BOT_TOKEN}/sendMessage`;
  const payload = {
    chat_id: env.TG_CHAT_ID,
    text,
    parse_mode: "HTML",
    link_preview_options:
      preview && url ? { url, prefer_large_media: true } : { is_disabled: true },
  };
  try {
    const r = await fetch(api, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!r.ok) console.error(`telegram HTTP ${r.status}: ${await r.text()}`);
    return r.ok;
  } catch (e) {
    console.error(`telegram error: ${e}`);
    return false;
  }
}

// ---------------------------------------------------------------------------
// Notifications
//
// Each notify* function returns the Set of product-id strings whose alert was
// actually delivered. For the summary path that's all-or-nothing (one message
// covers the whole batch); for the per-product path it's exactly the ones that
// succeeded. The caller records only delivered ids as seen.

async function notifyAdded(env, collection, label, products) {
  const base = storeBase(env);
  if (products.length > MAX_INDIVIDUAL) {
    const preview = products.slice(0, 5).map((p) => p.title).join(", ");
    const ok = await sendTelegram(env, {
      text:
        `🛍️ <b>${products.length} new in ${esc(label)}</b>\n` +
        `${esc(preview)} … and ${products.length - 5} more.\n` +
        `<a href="${collectionUrl(base, collection)}">View collection</a>`,
    });
    return ok ? new Set(products.map((p) => String(p.id))) : new Set();
  }
  const results = await Promise.all(
    products.map(async (p) => {
      const price = productPrice(p);
      const head = price ? `₹${esc(price)}` : "New product";
      const link = productUrl(base, p.handle || "");
      const ok = await sendTelegram(env, {
        text:
          `🛍️ <b><a href="${link}">${esc(p.title || "Product")}</a></b>\n` +
          `${head} · in ${esc(label)}`,
        url: link,
        preview: true,
      });
      return ok ? String(p.id) : null;
    })
  );
  return new Set(results.filter(Boolean));
}

// `items` is a list of { p, v }. Group by product so a product restocking
// several variants becomes one notification that names the variants.
async function notifyRestocked(env, collection, label, items) {
  const base = storeBase(env);
  const byProduct = new Map();
  for (const { p, v } of items) {
    if (!byProduct.has(p.id)) byProduct.set(p.id, { p, vs: [] });
    byProduct.get(p.id).vs.push(v);
  }
  const groups = [...byProduct.values()];

  if (groups.length > MAX_INDIVIDUAL) {
    const preview = groups.slice(0, 5).map((g) => g.p.title).join(", ");
    const ok = await sendTelegram(env, {
      text:
        `🔥 <b>${groups.length} back in stock in ${esc(label)}</b>\n` +
        `${esc(preview)} … and ${groups.length - 5} more.\n` +
        `<a href="${collectionUrl(base, collection)}">View collection</a>`,
    });
    return ok ? new Set(groups.map((g) => String(g.p.id))) : new Set();
  }

  const results = await Promise.all(
    groups.map(async ({ p, vs }) => {
      const names = vs.map(variantName).join(", ");
      const prices = vs.map((v) => v.price).filter(Boolean);
      let priceLine = "";
      if (prices.length) {
        const cheapest = prices.reduce((a, b) => (Number(b) < Number(a) ? b : a));
        priceLine = prices.length === 1 ? ` · ₹${cheapest}` : ` · from ₹${cheapest}`;
      }
      const link = productUrl(base, p.handle || "");
      const ok = await sendTelegram(env, {
        text:
          `🔥 <b><a href="${link}">${esc(p.title || "Product")}</a></b>\n` +
          `Back in stock: ${esc(names + priceLine)}\nin ${esc(label)}`,
        url: link,
        preview: true,
      });
      return ok ? String(p.id) : null;
    })
  );
  return new Set(results.filter(Boolean));
}

// ---------------------------------------------------------------------------
// Per-collection processing

async function processAdded(env, name, label, products) {
  const seen = await loadSeen(env, name);

  if (seen === null) {
    const currentIds = products.map((p) => String(p.id));
    await saveSeen(env, name, currentIds);
    await sendTelegram(env, {
      text:
        `👀 <b>Monitoring started: ${esc(label)}</b>\n` +
        `Now tracking ${currentIds.length} products. You'll be alerted on new additions.`,
    });
    return `[${name}] first run — seeded ${currentIds.length} products.`;
  }

  const fresh = products.filter((p) => !seen.has(String(p.id)));
  let delivered = new Set();
  if (fresh.length) delivered = await notifyAdded(env, name, label, fresh);

  // `seen` is MONOTONIC for new-arrivals: once a product id is recorded we keep
  // it forever and only ADD newly-delivered ids. We must NOT prune ids that are
  // merely absent from this fetch — the ~5k-item feed is paginated across ~20
  // pages of a live, changing collection, so boundary products intermittently
  // drop out of a fetch. Pruning on absence made them re-appear as "new" and
  // re-notify every few minutes (the duplicate-notifications bug).
  const newSeen = new Set(seen);
  for (const id of delivered) newSeen.add(id);
  if (!setsEqual(newSeen, seen)) await saveSeen(env, name, newSeen);

  const failed = fresh.length - delivered.size;
  return `[${name}] ${fresh.length} new (${delivered.size} delivered${failed ? `, ${failed} retrying` : ""}); ${products.length} in feed.`;
}

async function processRestocked(env, name, label, products) {
  const base = storeBase(env);
  // Candidate in-stock notifiable variants from the (cached) collection feed.
  const candidates = new Map(); // variant-id -> { p, v }  (in-stock + notifiable)
  const fetchedVariantIds = new Set(); // EVERY variant id seen in this fetch
  for (const p of products) {
    for (const v of p.variants || []) {
      fetchedVariantIds.add(String(v.id));
      if (v.available && isNotifiableVariant(v)) candidates.set(String(v.id), { p, v });
    }
  }

  // Health guard: this collection always has plenty in stock. If the feed
  // reports none (e.g. the `available` field vanished), don't wipe state.
  if (candidates.size === 0) {
    console.error(`[${name}] 0 in-stock notifiable variants — suspicious, skipping.`);
    return `[${name}] 0 in-stock notifiable variants — skipped.`;
  }

  const seen = await loadSeen(env, name);
  if (seen === null) {
    await saveSeen(env, name, [...candidates.keys()]);
    await sendTelegram(env, {
      text:
        `👀 <b>Monitoring started: ${esc(label)}</b>\n` +
        `Watching for restocks of testers, retail, partials & sets ` +
        `(decants, samples & miniatures ignored). Tracking ${candidates.size} ` +
        `in-stock variants — you'll be alerted when more come back.`,
    });
    return `[${name}] first run — seeded ${candidates.size} in-stock variants.`;
  }

  const freshIds = [...candidates.keys()].filter((id) => !seen.has(id));

  // Confirm candidate restocks in real time (.js), one fetch per product.
  const byHandle = new Map();
  for (const id of freshIds) {
    const { p } = candidates.get(id);
    if (!byHandle.has(p.handle)) byHandle.set(p.handle, []);
    byHandle.get(p.handle).push(id);
  }
  const confirmedIds = new Set();
  const confirmed = [];
  for (const [handle, ids] of byHandle) {
    const live = await liveAvailableVariants(base, handle);
    if (live === null) continue; // couldn't confirm — retry next run
    for (const id of ids) {
      if (live.has(id)) {
        confirmedIds.add(id);
        confirmed.push(candidates.get(id));
      }
    }
  }

  let delivered = new Set(); // product-id strings whose alert was delivered
  if (confirmed.length) delivered = await notifyRestocked(env, name, label, confirmed);

  // Rebuild seen. Keep a previously-seen variant if it's still an in-stock
  // candidate, OR if it's entirely ABSENT from this fetch — absence means a
  // pagination miss on this changing feed, not a sell-out, and dropping it
  // would make it re-alert as a "restock" when it reappears. Only drop a
  // variant that IS present in the fetch but is no longer an in-stock candidate
  // (a genuine sell-out) — so it correctly re-alerts when it actually restocks.
  // Newly confirmed variants are added only if their alert delivered.
  const newSeen = new Set();
  for (const id of seen) {
    if (candidates.has(id) || !fetchedVariantIds.has(id)) newSeen.add(id);
  }
  let deliveredVariants = 0;
  for (const id of confirmedIds) {
    if (delivered.has(String(candidates.get(id).p.id))) {
      newSeen.add(id);
      deliveredVariants++;
    }
  }
  if (!setsEqual(newSeen, seen)) await saveSeen(env, name, newSeen);

  return `[${name}] ${confirmed.length} confirmed restock(s) (${deliveredVariants} delivered); ${candidates.size} in stock; ${freshIds.length} candidate(s).`;
}

// ---------------------------------------------------------------------------
// Entry points

async function runChecks(env) {
  const base = storeBase(env);
  if (!base) {
    console.error("STORE_BASE_URL is not set; cannot fetch.");
    return "STORE_BASE_URL not set.";
  }
  const log = [];
  for (const c of COLLECTIONS) {
    const label = titleCase(c.name);
    let products;
    try {
      products = await fetchProducts(`${base}/collections/${c.name}/products.json`, c.maxPages);
    } catch (e) {
      log.push(`[${c.name}] fetch failed: ${e}`);
      console.error(log[log.length - 1]);
      continue;
    }
    if (products.length === 0) {
      log.push(`[${c.name}] 0 products — skipped (transient?).`);
      continue;
    }
    const fn = c.mode === "restocked" ? processRestocked : processAdded;
    log.push(await fn(env, c.name, label, products));
  }
  const out = log.join("\n");
  console.log(out);
  return out;
}

// Exported for the Node entry point (src/run.mjs) and tests.
export {
  isNotifiableVariant,
  processAdded,
  processRestocked,
  runChecks,
  setsEqual,
  COLLECTIONS,
};
